Add support for Keycloak

NotActuallyTerry opened the pull request:
This PR adds the ability to sync memberships from Keycloak groups to GitHub teams.

Right now, this only supports syncing from groups within a single Keycloak realm. If the need arises, I could look into adding support for defining the realm within the sync map.
As Keycloak supports using GitHub as an IdP, if you have this set up you can enable `KEYCLOAK_USE_GITHUB_IDP` to use their linked GitHub username directly. Further down the line it might be ideal to look at using the GitHub account ID in the sync.

NotActuallyTerry committed:
Add python-keycloak to Pipfile

NotActuallyTerry committed:
Add Keycloak integration

This integration allows Github Team Sync to pull group data from Keycloak via its admin REST API.
Users can choose to do a username sync as normal, or they can link to an existing Github account by enabling KEYCLOAK_USE_GITHUB_IDP. This requires the user to add Github as an Identity provider on the user's realm.

NotActuallyTerry committed:
Add documentation for Keycloak

primetheus commented:
Thanks for opening this up @NotActuallyTerry! I'll do some testing and get it merged in for a new release!

NotActuallyTerry committed:
Keycloak: define `members` as empty list

Defining members as None then adding group members to it appears to work in newer Python versions. However, this results in a TypeError in 3.9.
Changing this to define an empty list from the outset works better, and I probably should've done that to begin with

NotActuallyTerry commented:
No worries! I've pushed another commit fixing an issue that I ran into. Turns out my pipenv didn't fully override the system Python, so something that works in 3.11 but doesn't work in 3.9 slipped through :facepalm:

NotActuallyTerry committed:
Keycloak: do exact search for group

NotActuallyTerry committed:
Keycloak: raise Exception if group isn't found

NotActuallyTerry committed:
Keycloak: clarify settings for realms

NotActuallyTerry committed:
Keycloak: fix default realm handling

NotActuallyTerry committed:
Fix bug (also fix commit history later)

NotActuallyTerry committed:
Keycloak: Start page_start at 0

Turns out, you need to start your page at 0 to get all users!
How could this have happened!
